truffle/src/main/java/org/jruby/truffle/core/exception/ExceptionNodes.java

```diff
@@ -133,9 +133,23 @@ public AllocateNode(RubyContext context, SourceSection sourceSection) {
             allocateObjectNode = AllocateObjectNodeGen.create(context, sourceSection, null, null);
         }
 
-        @Specialization
+        @Specialization(guards = "!isNameError(rubyClass)")
         public DynamicObject allocate(DynamicObject rubyClass) {
-            return allocateObjectNode.allocate(rubyClass, nil(), null, nil(), null);
+            return allocateObjectNode.allocate(rubyClass, nil(), null, nil());
+        }
+
+        @Specialization(guards = "isNameError(rubyClass)")
+        public DynamicObject allocateNameError(DynamicObject rubyClass) {
+            return allocateObjectNode.allocate(rubyClass, nil(), null, null, nil());
+        }
+
+        public static boolean isNameError(DynamicObject rubyClass){
+           for(DynamicObject ancestor : Layouts.MODULE.getFields(rubyClass).ancestors()){
+                  if(Layouts.MODULE.getFields(ancestor).getName().equals("NameError")){
```

Comment:
I know you asked me about this and I don't think I fully understood what you were doing at the time. If you really have the class handle and you're looking for a direct class match (rather an "is a" relationship), you should be able to do a reference comparison to `coreLibrary().getNameErrorClass()`.